### PR TITLE
Fix property of type Node duplication when script is attached to a child node

### PR DESCRIPTION
**Pull request by warriormaster12**
This is a continuation of a last PR where I seemed to have only partially solved the problem with node duplication as pointed out by one of the comments [after the PR got merged](https://github.com/godotengine/godot/pull/83343#issuecomment-1872988602). 

This PR should hopefully solve the edge case when a script is attached to a child node but the user duplicates that node's parent.

### CODE REVIEW

File: scene/main/node.cpp [hunk 1]
```diff
@@ -2620,7 +2619,6 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
 			parent->move_child(dup, pos);
 		}
 	}
-
 	for (List<const Node *>::Element *N = node_tree.front(); N; N = N->next()) {
```
Comment (AThousandShips) [inline scene/main/node.cpp:2622]:
```suggestion

	for (List<const Node *>::Element *N = node_tree.front(); N; N = N->next()) {
```
Unnecessary and unrelated whitespace change

File: scene/main/node.cpp [hunk 2]
```diff
@@ -2559,7 +2559,6 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
 	if (get_name() != String()) {
 		node->set_name(get_name());
 	}
-
 #ifdef TOOLS_ENABLED
```
Comment (AThousandShips) [inline scene/main/node.cpp:2562]:
```suggestion

#ifdef TOOLS_ENABLED
```

File: scene/main/node.h [hunk 3]
```diff
@@ -237,6 +237,7 @@ class Node : public Object {
 	void _propagate_groups_dirty();
 	Array _get_node_and_resource(const NodePath &p_path);
 
+	void _duplicate_deferred_properties(const Node *p_root, const Node *p_original, Node *p_copy) const;
```
[the remark this answers is not in the document]
Comment (warriormaster12) [reply scene/main/node.h:240]: Not sure if it’s even a correct term to be used but I used property_utils.cpp as an example https://github.com/godotengine/godot/blob/0bcc0e92b3f0ac57d4c4650722f347593a258572/scene/property_utils.cpp#L77 

Maybe it should be called ```_duplicate_proprerties_node``` instead 🤔

[the remark this answers is not in the document]
Comment (AThousandShips) [reply scene/main/node.h:240]: That one refers to properties that are deferred afaik, specifically, unsure what to name this one myself

**Comment by warriormaster12** [comment]
@AThousandShips is it possible to get this cherry-picked for 4.2?

**Comment by AThousandShips** [comment]
I don't know the area well enough to make that call

**Comment by warriormaster12** [comment]
Can you confirm that this doesn’t happen with a version of godot before my PRs?

File: scene/main/node.cpp [hunk 4]
```diff
@@ -2654,25 +2654,9 @@ Node *Node::_duplicate(int p_flags, HashMap<const Node *, Node *> *r_duplimap) c
...
-					if (arr.get_typed_builtin() == Variant::OBJECT) {
-						for (int i = 0; i < arr.size(); i++) {
-							Node *property_node = Object::cast_to<Node>(arr[i]);
-							if (property_node && is_ancestor_of(property_node)) {
-								arr[i] = current_node->get_node_or_null(get_path_to(property_node));
-							}
-						}
-						value = arr;
-					}
+				if (value.get_type() != Variant::OBJECT && !(value.get_type() == Variant::ARRAY && static_cast<Array>(value).get_typed_builtin() == Variant::OBJECT)) {
```
Comment (AThousandShips) [inline scene/main/node.cpp:2657]:
```suggestion
				if (value.get_type() != Variant::OBJECT && (value.get_type() != Variant::ARRAY || static_cast<Array>(value).get_typed_builtin() != Variant::OBJECT)) {
```

File: scene/main/node.cpp [hunk 5]
```diff
@@ -2761,6 +2749,48 @@ void Node::remap_nested_resources(Ref<Resource> p_resource, const HashMap<Ref<Re
 }
 #endif
 
+// Duplicates properties that store a Node.
+// This has to be called after nodes have been duplicated since
+// we only then get a full picture of how the duplicated node tree looks like
```
Comment (AThousandShips) [inline scene/main/node.cpp:2754]:
```suggestion
// only then do we get a full picture of how the duplicated node tree looks like.
```

File: scene/main/node.cpp [hunk 6]
```diff
@@ -2761,6 +2749,48 @@ void Node::remap_nested_resources(Ref<Resource> p_resource, const HashMap<Ref<Re
...
+			if (property_node && (p_root->is_ancestor_of(property_node) || property_node->get_path_to(p_root) == ".")) {
+				value = p_copy->get_node_or_null(p_original->get_path_to(property_node));
+				p_copy->set(name, value);
+			}
+		} else if (value.get_type() == Variant::ARRAY) {
+			Array arr = value;
+			if (arr.get_typed_builtin() == Variant::OBJECT) {
+				for (int i = 0; i < arr.size(); i++) {
+					Node *property_node = Object::cast_to<Node>(arr[i]);
+					if (property_node && (p_root->is_ancestor_of(property_node) || property_node->get_path_to(p_root) == ".")) {
```
Comment (AThousandShips) [inline scene/main/node.cpp:2779]:
```suggestion
					if (property_node && (p_root == property_node || p_root->is_ancestor_of(property_node))) {
```

File: scene/main/node.cpp [hunk 7]
```diff
@@ -2761,6 +2749,48 @@ void Node::remap_nested_resources(Ref<Resource> p_resource, const HashMap<Ref<Re
...
+			continue;
+		}
+		String name = E.name;
+		if (name == script_property_name) {
+			continue;
+		}
+		Variant value = p_original->get(name);
+		if (value.get_type() == Variant::OBJECT) {
+			Node *property_node = Object::cast_to<Node>(value);
+			if (property_node && (p_root->is_ancestor_of(property_node) || property_node->get_path_to(p_root) == ".")) {
```
Comment (AThousandShips) [inline scene/main/node.cpp:2770]:
```suggestion
			if (property_node && (p_root == property_node || p_root->is_ancestor_of(property_node))) {
```

File: scene/main/node.cpp [hunk 8]
```diff
@@ -2844,6 +2832,48 @@ void Node::remap_nested_resources(Ref<Resource> p_resource, const HashMap<Ref<Re
...
+void Node::_duplicate_properties_node(const Node *p_root, const Node *p_original, Node *p_copy) const {
+	List<PropertyInfo> props;
+	p_copy->get_property_list(&props);
+	StringName script_property_name = CoreStringNames::get_singleton()->_script;
+	for (const PropertyInfo &E : props) {
+		if (!(E.usage & PROPERTY_USAGE_STORAGE)) {
+			continue;
+		}
+		String name = E.name;
+		if (name == script_property_name) {
```
Comment (KoBeWi) [inline scene/main/node.cpp:2847]: I think this check is irrelevant, because you are casting to Node, so a Script won't pass anyway.

**Comment by KoBeWi** [comment]
There is some bug where if you duplicate a node and the duplicated node has its properties updated, the original will also have them changed.

https://github.com/godotengine/godot/assets/2223172/634e896f-e959-442e-a468-110e2b91f663

**Comment by warriormaster12** [comment]
@KoBeWi Try again, value now is deeply duplicated so the bug shouldn't happen.

File: scene/main/node.cpp (hunk 8 above)
Comment (warriormaster12) [reply scene/main/node.cpp]: Removed the if statement

Comment (KoBeWi) [reply scene/main/node.cpp]: Now `script_property_name` is unused. You should remove it too.

**Review by KoBeWi — APPROVED** [review]

File: scene/main/node.cpp (hunk 8 above)
Comment (warriormaster12) [reply scene/main/node.cpp]: Done, how silly of me

**Comment by akien-mga** [comment]
Thanks!
